Added support for getting the cover art from mpd

eternalfrustation opened the pull request:
As proposed in the previous PR (#466), this PR implements displaying cover art from mpd using the readpicture command

eternalfrustation committed:
Added support for using a cover image which has a specific static path in the music module

eternalfrustation committed:
Added support for using kbps, kBps for total network usage

eternalfrustation committed:
Added support for getting the cover art from mpd

eternalfrustation committed:
Satisfying Clippy

JakeStanger commented on src/clients/music/mpd.rs:
```diff
@@ -119,6 +122,69 @@ impl Client {
+            TypedResponseError::missing("cover art")
+        })?;
+        offset += slice.1.len();
+        buffer.write_all(slice.1.as_slice()).unwrap();
```
You should avoid using `unwrap` in production code. Generally errors should be properly handled, logging them and/or performing any required logic to catch it. 

Where you are absolutely certain that a call will not fail under any normal circumstance, or that the error is severe enough to warrant exiting, you should use `expect` instead to make it explicit what we're expecting to (not) happen.

JakeStanger reviewed (changes requested):
Thanks for taking the time to rewrite this. There are some issues I've raised with your implementation which will need addressing, but it's mostly just simplifying things and refactoring.

In future, can you please have a read of the [contributing guide](https://github.com/JakeStanger/ironbar/blob/master/CONTRIBUTING.md) - it's very short but has a few requirements I ask people to stick to in order to make both our lives easier.

The build is also failing currently, which will need to be addressed of course.

eternalfrustation commented on src/clients/music/mod.rs:
```diff
@@ -31,7 +31,7 @@ pub struct Track {
     pub genre: Option<String>,
     pub track: Option<u64>,
-    pub cover_path: Option<String>,
+    pub cover_image: Option<image::DynamicImage>,
```
Mpd provides us with an encoded image, usually as a jpeg, but could also be an Png or something else. As far as i could see, the most obvious way to handle that was by decoding the image ourselves and then passing the decoded pixels to Pixbuf.

JakeStanger replied:
GTK can handle various file types automatically, I don't think there's any need to handle that in our application code. Have you tested passing the bytes directly?

Creating an image from the bytes as I've shown above (specially `ImageProvider::create_and_load_surface(&pixbuf, &image)`) ensures the image loads correctly in different scaling settings.

eternalfrustation committed:
feat: Modifying the image via message passing

As suggested previously, This commit uses message passing
to send the image from the client to the main thread
while retaining the original way of setting the album art.

eternalfrustation committed:
fix: Removed the path section from Cargo.toml

Removed the patch section for ahash from Cargo.toml

eternalfrustation committed:
fix: Ran cargo update
Due to the previous commit, cargo could not find the
proper ahash version, ran cargo update to fix that.

eternalfrustation committed:
fix: Removed the image crate from dependencies

Now using gdk's image parser for parsing the mpd's
thumbnails that are read from readpicture command

eternalfrustation commented:
I have done all the requested changes

eternalfrustation commented:
What kind of Clippy rules are you using

eternalfrustation committed:
fix: Moved the Cacellable and MemoryStream imports out of the cfg_if block

JakeStanger reviewed (changes requested):
Looking a lot better so thanks for the work, there's still a few changes that need to be made. Once those are done I'll give it another review with a fine comb and we should be about there.

JakeStanger commented:
> What kind of Clippy rules are you using

The defaults: https://github.com/JakeStanger/ironbar/blob/3df01ea93a27ba80e2be8c3fe832903c3e7871d2/.github/workflows/build.yml#L61

eternalfrustation commented on src/clients/music/mpd.rs:
```diff
@@ -93,11 +96,17 @@ impl Client {
-            send!(tx, update);
+            let update_info = PlayerUpdate::Update(Box::new(track), status);
+            let update_image = PlayerUpdate::UpdateImage(cover_image);
+            send!(tx, update_image);
```
Is there any way to do send the signal from the popup? That would make it a lot easier.

eternalfrustation replied:
Or getting a reference to MusicPlayer, or PersistentClient, or anything from the Popup

JakeStanger replied:
The controller should be responsible for communicating between the client and the popup. Since the controller is already listening to the client events, you should be able to detect the necessary change there and just call a function on the client to get the image.

eternalfrustation committed:
fix: Only sending the GetAlbumArt signal when needed

Only sending the GetAlbumArt signal when track changes and
the cover_art path does not exist.

eternalfrustation committed:
feat: Satisfying Clippy

JakeStanger commented:
Please fix the merge conflict before I review

eternalfrustation committed:
feat: Removed Cargo.lock for solving merge conflict

eternalfrustation committed:
chore: Updating the branch to master

eternalfrustation committed:
chore: Updating the branch to master

eternalfrustation committed:
chore: Removing the cargo.toml again

JakeStanger commented:
Thanks that'd be great if you can. This PR was most of the way there, so it probably doesn't need too much more than rebasing and a quick modernisation pass

Diax170 commented:
for now you can use mpd-mpris and change the bar settings to use mpris, cover works here
